app/services/record_declarations/base.rb

```diff
@@ -106,9 +96,56 @@ def create_record!
       end
     end
 
+    def valid_courses_for_user
+      valid_courses = []
+      valid_courses << "ecf-mentor" if user.mentor?
+      valid_courses << "ecf-induction" if user.early_career_teacher?
+      valid_courses += NPQCourse.all.map(&:identifier) if user.npq?
+      valid_courses
+    end
+
+    def profile_exists
+      return if errors.any?
+
+      unless user_profile
+        errors.add(:user_profile, I18n.t(:invalid_participant))
+      end
+    end
+
+    def date_has_the_right_format
+      return if declaration_date.blank?
+
+      errors.add(:declaration_date, I18n.t(:invalid_declaration_date)) unless declaration_date.match(RFC3339_DATE_REGEX)
+      errors.add(:declaration_date, I18n.t(:future_declaration_date)) if parsed_date > Time.zone.now
+    rescue StandardError
+      errors.add(:declaration_date, I18n.t(:invalid_declaration_date))
+    end
+
     def validate_provider!
       # TODO: Remove the nil? check and fix the test setup so that they build the school cohort, partnership and give us back the actual lead_provider.
       raise ActionController::ParameterMissing, I18n.t(:invalid_participant) unless actual_lead_provider.nil? || lead_provider_from_token == actual_lead_provider
     end
+
+    def validate_milestone!
+      unless next_milestone.start_date < parsed_date
+        raise ActionController::ParameterMissing, I18n.t(:declaration_before_milestone_start)
+      end
+
+      unless next_milestone.milestone_date > parsed_date
+        raise ActionController::ParameterMissing, I18n.t(:declaration_after_milestone_cutoff)
+      end
+    end
+
+    def next_milestone
+      unless schedule
+        raise ActionController::ParameterMissing, I18n.t(:schedule_missing)
+      end
+
+      if participant_declarations.count >= schedule.milestones.count
+        raise ActionController::ParameterMissing, I18n.t(:too_many_declarations)
+      end
```

Comment:
What's this? There's no business decision on restricting the number of participant_declarations yet.

Reply:
This is what protects us from out of bound exception.

Reply:
Out of bounds how? That's really nothing to do with milestone dates and doesn't seem to be a good fit for this PR.

Reply:
There are like 5 or 6 milestones per schedule. To get the milestone we check against this implementation says 'count the existing declarations, assume they are 1-1 to the milestones, and so pick the milestone in the ordered list of milestones that is next'. And what happens when there is no next and we receive a declaration?

I don't expect this solution will stay in forever - one other thing we want to validate, which will probably superseed this, is that the declaration type is right for the milestone in question. But until we do that, this is a simple way of fetching a milestone and showing that date validations work for this milestone.